### ISSUE TITLE
Reproducing pytorch transforms values with DALI

### ISSUE THREAD

**Issue by tarunn2799** [NVIDIA/DALI#2789]
Hey, so I'm currently upgrading an existing system that uses Torchvision for augmentations, to use DALI for our data pipelines. We've finished integration, and achieved much faster training. 
My current pipeline uses DALI for training and pytorch for validation and testing. 

I have been trying to see if the values generated by passing an image through DALI transforms match those when it is passed through the exact same Torchvision counterparts. 
I notice a non-negligible difference between the two, and I am not sure why this is happening.  The difference grows larger with every successive transform I add. 
Currently my pipeline only contains an ImageDecoder, Resize, and Normalization in that order.

![image](https://user-images.githubusercontent.com/28048963/110935859-fcb34d00-8355-11eb-868c-cd1671f83bcb.png)

I ran a test to see if the output tensors match, and this was what I got. This is just with 1 decimal point taken into account. It reaches 87% mismatched elements when I make it 2.  What could be causing this difference?

**Comment by JanuszL** [comment]
Hi,
The biggest source of discrepancy is the ImageDecoder. You can even compare the CPU and Mixed backends. The reason for this difference is the fact there is no gold standard for JPED decoding, and there is some freedom for interpolation and filtering.
Regarding Resize I would double-check what interpolation is used by default by Torchvision and use a corresponding one from DALI (I think it uses PIL under the hood and bilinear, in that case, may mean TRIANGULAR in DALI).
Normalization should be mathematically accurate (up to the floating-point math that may differ between CPU and GPU).

**Comment by mzient** [comment]
By default, DALI's resize uses triangular interpolation (which is equivalent to bilinear for _upscaling_, but it features antialiasing when downscaling). Torchvision, by default, uses bilinear interpolation (which suffers from aliasing when downscaling).
As @JanuszL said, there's no JPEG decoding standard - in general, the better the PSNR for encode-decode, the better, so different decoders employ different tricks to improve the result - sometimes optimized for images from specific field.
Moreover, when using non-explicit output sizes (modes other than "stretch"), then DALI uses subpixel ROI adjustment to maintain aspect ratio at the cost of slightly misplaced pixel centers near the edges. This behavior can be disabled by setting `subpixel_scale` argument to `False`.
Also, I think that default torchvision produces `float` output by default whereas DALI's `resize` produces images of the same type as the input - that, too, can be changed.
Please refer to the extensive tutorials on resize and interpolation methods in DALI and the Resize operator's reference
https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/image_processing/resize.html
https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/image_processing/interp_types.html
https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.Resize

**Comment by tarunn2799** [comment]
Thank you!
